packages/loopring_v3/contracts/impl/libexchange/ExchangeBlocks.sol

```diff
@@ -374,50 +263,238 @@ library ExchangeBlocks
                 require(inputEndingHash == endingHash, "INVALID_ENDING_HASH");
                 numWithdrawalRequestsCommitted += uint32(count);
             }
-        } else if (
-            blockType != ExchangeData.BlockType.OFFCHAIN_WITHDRAWAL &&
-            blockType != ExchangeData.BlockType.ORDER_CANCELLATION &&
-            blockType != ExchangeData.BlockType.TRANSFER) {
+        } else if (_block.blockType == ExchangeData.BlockType.TRANSFER) {
+            validateConditionalTransfers(
+                S,
+                _block.data,
+                _block.auxiliaryData
+            );
+        } else if (_block.blockType != ExchangeData.BlockType.OFFCHAIN_WITHDRAWAL) {
             revert("UNSUPPORTED_BLOCK_TYPE");
         }
 
-        // Hash all the public data to a single value which is used as the input for the circuit
-        bytes32 publicDataHash = data.fastSHA256();
-
-        // Store the approved withdrawal data onchain
-        bytes memory withdrawals = new bytes(0);
-        if (blockType == ExchangeData.BlockType.ONCHAIN_WITHDRAWAL ||
-            blockType == ExchangeData.BlockType.OFFCHAIN_WITHDRAWAL) {
+        // Process withdrawals
+        if (_block.blockType == ExchangeData.BlockType.ONCHAIN_WITHDRAWAL ||
+            _block.blockType == ExchangeData.BlockType.OFFCHAIN_WITHDRAWAL) {
             uint start = 4 + 32 + 32;
-            if (blockType == ExchangeData.BlockType.ONCHAIN_WITHDRAWAL) {
+            if (_block.blockType == ExchangeData.BlockType.ONCHAIN_WITHDRAWAL) {
                 start += 32 + 32 + 4 + 4;
             }
-            uint length = 7 * blockSize;
+            uint length = 7 * _block.blockSize;
+            bytes memory withdrawals = new bytes(0);
             assembly {
                 withdrawals := add(data, start)
                 mstore(withdrawals, length)
             }
+            S.distributeWithdrawals(withdrawals);
```

Comment:
distributeWithdrawals may take a lot of gas fees which will prevent us from enabling large withdrawal blocks in the future. 

Maybe we can do the following:
1. move `S.distributeWithdrawals(withdrawals);`to the very end of this method;
1. inside `distributeWithdrawals`, we calculate the gas remains, and if it is enough for calling `transferTokens` for the next withdrawal in the block, we do it; otherwise, we skip calling transferTokens, but simply change `amountWithdrawable`

Reply:
If we do this, it provides operators a way to enable larger blocks for withdrawals - he can simply calculate the right amount of gas limit for such a block so that the code will not spend gas on any `transferToken` invitation. Then the operator can use another tx to withdraw funds to users by calling `withdrawFromApprovedWithdrawal`.

Speaking of `withdrawFromApprovedWithdrawal`, maybe we can provide a method for batch withdrawal so operators can do that for better user experience at his own cost.

Reply:
@Brechtpd 
> Some quick calculations to see how this effects our throughput (best case scenario). As you can see even quite small blocks aren't too far from the best case with the biggest block sizes, so maybe for our next trusted setup we can do just up to 512 trades/block (or even 256 trades/block).

If the above statement of your still holds true, maybe we don't have to worry about what I worried - supporting of very large blocks. 512 seems to be good size. Large blocks do consume expeisve hardware and have a longer delay.

Reply:
It seems my previous comments are not valid, as we do commitBlock first then do verifyBlocks later, it's not easy to calculate the remaining gas for distribution withdrawals.

Reply:
Using `amountWithdrawable` to store the withdrawal data will actually be MORE expensive most times than just doing the withdrawal. If `amountWithdrawable` is 0 (which will be most times) it costs 20,000 gas to store the data, while doing the transfer immediately will use something like 15,000 gas on average for just doing the transfer immediately.

I wouldn't worry too much about being able to use large blocks for withdrawals because it will be expensive no matter what. The cost for the block + verification itself is tiny compared to the actual transfers.

> Speaking of withdrawFromApprovedWithdrawal, maybe we can provide a method for batch withdrawal so operators can do that for better user experience at his own cost.

Because the transfers will happen automatically I don't think it's likely it will be used a lot. But anyone can call the function so any contract can be implemented to do this batching if needed.